labs/optimizely-data-cli/oevents

```diff
@@ -0,0 +1,651 @@
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-03/experiment=56789/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-04/experiment=56789/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-05/experiment=56789/
+"
+
+################################################################################
+# Logging and Error output                                                     #
+################################################################################
+
+# VerboseEcho
+# Echo the supplied text if the --verbose flag is on
+function VerboseEcho()
+{
+  if [ "$verbose" = true ]; then 
+    echo "$@"
+  fi
+}
+
+# ErrorEcho
+# Echo the supplied text to stderr 
+function ErrorEcho()
+{
+  echo "$@" >&2
+}
+
+################################################################################
+# Prerequisites                                                                #
+################################################################################
+
+function Require()
+{
+  local requirement="$1"
+
+  if ! command -v "$requirement" &> /dev/null; then
+    ErrorEcho "Error: $CLI_NAME requires $requirement to run.  Please install it and try again."
+    exit 1
+  fi
+}
+
+
+################################################################################
+# Date and Time                                                                #
+################################################################################
+
+# TimestampMsToDateString
+# Convert a unix timestamp (in milliseconds) to a date string and return the result
+function TimestampMsToDateString()
+{
+  local timestamp_ms="$1"
+  local timestamp_s=$(( timestamp_ms / 1000 ))
+  if [[ $(uname) = 'Darwin' ]]; then
+    echo $(date -r "$timestamp_s")
+  else
+    echo $(date -d "@$timestamp_s")
+  fi
+}
+
+# IncrDay
+# Increment a YYYY-MM-DD date string by one day and return the result
+function IncrDay()
+{
+  local to_be_incremented="$1"
+  
+  # From https://stackoverflow.com/questions/25701265/how-to-generate-a-list-of-all-dates-in-a-range-using-the-tools-available-in-bash
+  if [[ $(uname) = 'Darwin' ]]; then
+    echo $(date -j -v+1d -f "%Y-%m-%d" $to_be_incremented "+%Y-%m-%d")
+  else
+    echo $(date -I -d "$to_be_incremented + 1 day")
+  fi
+}
+
+# AssertIsBeforeOrEqual
+# Given two dates, assert that the first is before or equal to the second
+function AssertIsBeforeOrEqual()
+{
+  d1="$1"
+  d2="$2"
+
+  if [[ $(uname) = 'Darwin' ]]; then
+    ts1=$(date -j -u -f "%Y-%m-%d" "$d1" "+%s")
+    ts2=$(date -j -u -f "%Y-%m-%d" "$d2" "+%s")
+  else
+    ts1=$(date -d "$d1" +"%s")
+    ts2=$(date -d "$d2" +"%s")
+  fi
+
+  if [ "$ts1" -gt "$ts2" ]; then
+    ErrorEcho "Error: invalid date range: $d1 - $d2"
+    exit 1
+  fi
+}
+
+# ComputeDateRange
+# Compute a date range between $start and $end, and store the result in $date_range
+function ComputeDateRange()
+{
+  date_range=()
```

Comment:
Regular arrays can be constructed and assigned without `-a`, it's personal preference afaik. In BASH4+, associative arrays _require_ the `-A` flag but regular ones, again, can be constructed directly and I think this might be a POSIX thing or a BASHism in either direction, I just can't remember. Either way, this will work if the intent is for `date_range` to be an INDEXED ARRAY.

Reply:
`date_range` is a global variable (computed from the `--start` and `--end` command line arguments).  I changed `date_range`, `s3_relative_paths`, and `s3_absolute_paths` to SCREAMING_CASE to better communicate this, per another one of your recommendations.